core/arch/arm/sta/interrupt_tests.c

```diff
@@ -0,0 +1,264 @@
+ * this list of conditions and the following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ * this list of conditions and the following disclaimer in the documentation
+ * and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+#include <stdio.h>
+#include <trace.h>
+#include <kernel/static_ta.h>
+#include <kernel/interrupt.h>
+#include <kernel/misc.h>
+#include <kernel/tee_time.h>
+#include <string.h>
+
+#define TA_NAME		"interrupt_tests.ta"
+
+#define INTERRUPT_TESTS_UUID \
+		{ 0x48d58475, 0x3d5e, 0x4202, \
+		{ 0xa7, 0x75, 0x97, 0x85, 0xd2, 0x0f, 0x78, 0xae } }
+
+#define CMD_INTERRUPT_TESTS	0
+
+#define SGI_NUM		16
+#define PPI_NUM		32
+
+#define TEST_SGI_ID	11
+#define TEST_PPI_ID	29
+#define TEST_SPI_ID	61
+
+/*
+ * Trusted Application Entry Points
+ */
+
+static TEE_Result create_ta(void)
+{
+	return TEE_SUCCESS;
+}
+
+static void destroy_ta(void)
+{
+}
+
+static TEE_Result open_session(uint32_t ptype __unused,
+			       TEE_Param params[4] __unused,
+			       void **ppsess __unused)
+{
+	return TEE_SUCCESS;
+}
+
+static void close_session(void *psess __unused)
+{
+}
+
+static size_t test_sgi_value[8];
+static size_t test_spi_value[8];
+static size_t test_ppi_value[8];
+static size_t expect_sgi_value[8];
+static size_t expect_spi_value[8];
+static size_t expect_ppi_value[8];
+
+static enum itr_return __maybe_unused ihandler_ok(struct itr_handler *handler)
+{
+	size_t core_num = get_core_pos();
+
+	assert(core_num < 8);
+
+	if (handler->it < SGI_NUM)
+		test_sgi_value[core_num]++;
+	else if (handler->it < PPI_NUM)
+		test_ppi_value[core_num]++;
+	else
+		test_spi_value[core_num]++;
+
+	DMSG("I'm core [%zu].\n", core_num);
+	DMSG("I have handled the interrupt[%zu].\n", handler->it);
+	return ITRR_HANDLED;
+}
+
+struct itr_handler sgi_handler = {
+	.it = TEST_SGI_ID,
+	.handler = ihandler_ok,
+};
+
+struct itr_handler spi_handler = {
+	.it = TEST_SPI_ID,
+	.handler = ihandler_ok,
+};
+
+struct itr_handler ppi_handler = {
+	.it = TEST_PPI_ID,
+	.handler = ihandler_ok,
+};
+
+static TEE_Result test_sgi(void)
+{
+	itr_add(&sgi_handler);
+	itr_enable(TEST_SGI_ID);
+
```

Comment:
also use `CFG_TEE_CORE_NB_CORE` here so that code is portable across platforms with less than 4 cores.